Jump to PLA temps in LCD temperature settings when starting from zero

wavexx opened the pull request:
When turning on one heater manually through the Settings -> Temperature menu, it's painfully slow to turn encoder to reach _any_ useful temperature (usually, a minimum of 180C is required).

This PR allows to set an initial "jump" value when starting from zero. It's used in the Settings -> Temperature menus for both the Nozzle and Bed, setting the jump point to 215/60C respectively.

That is, when the nozzle temp is zero, and the user is trying to increase the value, immediately set the value to 215 for further editing. The jump works only once after entering the menu, and the user is still allowed to decrease the value after the jump, so there's no functional restriction. For this reason, 215/60C seem a pretty good all-around value, since most other materials require higher temperatures.

This fixes #3085

It is somewhat expensive in terms of code size, increasing flash usage by 248 bytes (sadly, most of which is due to the ugly menu call interface).

wavexx committed:
Jump to PLA temps in LCD settings when starting from zero

Allow the LCD menu update function to preset an initial value during the
first encoder increase from the minimal (usually zero) value.

This is useful to jump to a more sensible initial value when turning on
an heater which is currently disabled. The user is still allowed to
decrease the value after the jump, so there's no functional restriction.

3d-gussner reviewed (commented):
I would suggest to jump to `EXTRUDE_MINTEMP` as this is close to the PLA temp but also to cold pull temp or MBL without oosing.

wavexx commented:
The temperature for a cold pull is not entirely defined though, depends on the material. MINTEMP is probably already too high for a cold pull.

I usually pull starting from cold, so any temp would do.

But I'm open to this. @leptun @gudnimg ?

leptun commented:
I'd jump to around 200C. This is usually the minimum temperature that is really used with any material. Jumping to EXTRUDE_MINTEMP is dangerous since due to temperature oscillations, the cold extrude check will kick in. The PLA temp I think is a bit too high and it might cause PLA to ooze, I'd much rather start at a safer temperature where the plastc oozes very little.

wavexx commented:
200 sounds quite reasonable to me.

wavexx commented:
The biggest problem here with the code size is that menu_item_edit_P is a template which gets expanded for uint8_t and int16_t (depending on the menu). This in turns generates an almost complete copy of _menu_edit_P.

IMHO these two functions could take a void* pointer and check for the type internally. And/or be rewritten so that most of the code just works with int16_t, and just does a final cast at the end.

The 8 and 16 bit types have different rendering. menu_draw_P have two completely different functions depending on type, which is also bad style to abuse template specialization for this.

wavexx committed:
Make temperature jumping configurable in the variant file

Use this new feature in the Settings -> Temperature menus.

wavexx commented:
I changed the default temperature to 200 and made it configurable in the variant file. I also added the same for the fan speed (where the default is simply 50%).

3d-gussner reviewed (approved):
Tested on MK404 and real printer
- Jump to min value :heavy_check_mark: 
  - Settings -> Temperature -> Nozzle :heavy_check_mark: 
  - Settings -> Temperature -> Bed :heavy_check_mark: 
  - Settings -> Temperature -> Fan :heavy_check_mark:

- Turning the knob clockwise jumps to set value
- Turning the knob counter clockwise disables the jump and following clockwise turns increase the temp as known.

3d-gussner commented:
Cost +250 bytes of flash

wavexx committed:
Optimize menu editing by reducing template duplication

Avoid instantiating multiple copies of menu_item_edit_P/_menu_edit_P:
perform type switching at runtime.

gudnimg commented on Firmware/menu.h:
```diff
@@ -20,7 +20,8 @@ typedef struct
     const char* editLabel;
-    void* editValue;
+    uint8_t editValueBits; // 8 or 16
+    void* editValuePtr;
```
I like this rename 👍 makes it more clear its a pointer to some other variable

wavexx commented:
Should hopefully be less wasteful with the latest changes.

3d-gussner reviewed (approved):
Retested on MK404

Save 92 bytes

gudnimg reviewed (approved):
Works great on my MK3S+, everything seems to work as expected.
